Add support for `:include` index option

steve-abrams opened the pull request:
### Summary

This adds the option `:include` to PostgreSQL btree indexes to include non-key columns in the index using the `INCLUDE` parameter.

The `INCLUDE` parameter of `CREATE INDEX` is described in the [PostgreSQL docs](https://www.postgresql.org/docs/current/sql-createindex.html):

> The optional INCLUDE clause specifies a list of columns which will be included in the index as non-key columns.

Example:

```ruby
# In a migration
def change
  add_index :users, :email, include: :id
end
```

will result in:

```sql
CREATE INDEX index_users_on_email USING btree (email) INCLUDE (id)
```

Example with multiple columns:

```ruby
# In a migration
def change
  add_index :users, :email, include: [:id, :created_at]
end
```

will result in:

```sql
CREATE INDEX index_users_on_email USING btree (email) INCLUDE (id, created_at)
```

### Other Information

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails! -->

<!--
Note: Please avoid making *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.
Create a pull request when it is ready for review and feedback
from the Rails team :).
-->

steve-abrams committed:
Add support for include index option

Add support for including non-key columns in
btree indexes for PostgreSQL with the INCLUDE
parameter.

Example:

    def change
      add_index :users, :email, include: [:id, :created_at]
    end

Will result in:

    CREATE INDEX index_users_on_email USING btree (email) INCLUDE (id,
    created_at)

The INCLUDE parameter is described in the PostgreSQL docs:
  https://www.postgresql.org/docs/current/sql-createindex.html

steve-abrams commented:
👋 Hello! This is my first time contributing to Rails. I'm a little uncertain if I missed some tests that need updating, so please do point me in the right direction if I missed anything.

eileencodes commented on activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb:
```diff
@@ -101,6 +102,7 @@ def visit_CreateIndexDefinition(o)
           sql << "USING #{index.using}" if supports_index_using? && index.using
           sql << "(#{quoted_columns(index)})"
           sql << "WHERE #{index.where}" if supports_partial_index? && index.where
+          sql << "INCLUDE (#{quoted_include_columns(index.include)})" if supports_index_include? && index.include
```
Is it possible to handle this in the postgres schema_creation file (activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb) instead since this is a postgres only feature?

steve-abrams replied:
Yes, I've moved it and the associated methods to the postgres `schema_creation`  file. I also moved `quoted_include_columns_for_index` to the postgres `schema_statements` file.

eileencodes commented on activerecord/lib/active_record/connection_adapters/abstract/schema_creation.rb:
```diff
@@ -101,6 +102,7 @@ def visit_CreateIndexDefinition(o)
           sql << "USING #{index.using}" if supports_index_using? && index.using
           sql << "(#{quoted_columns(index)})"
           sql << "WHERE #{index.where}" if supports_partial_index? && index.where
+          sql << "INCLUDE (#{quoted_include_columns(index.include)})" if supports_index_include? && index.include
```
Is it possible to handle this in the postgres schema_creation file (activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb) instead since this is a postgres only feature?

eileencodes commented on activerecord/test/schema/schema.rb:
```diff
@@ -275,6 +275,7 @@
     t.index [:name, :description], length: 10
     t.index [:firm_id, :type, :rating], name: "company_index", length: { type: 10 }, order: { rating: :desc }
     t.index [:firm_id, :type], name: "company_partial_index", where: "(rating > 10)"
+    t.index [:firm_id, :type], name: "company_include_index", include: [:name, :account_id]
```
There's a postgres specific schema in activerecord/test/schema/postgresql_specific_schema.rb which might be a more appropriate file for this change since this won't work on any other adapters.

steve-abrams replied:
👍 I moved the index to `activerecord/test/schema/postgresql_specific_schema.rb`.

eileencodes commented on activerecord/test/schema/schema.rb:
```diff
@@ -275,6 +275,7 @@
     t.index [:name, :description], length: 10
     t.index [:firm_id, :type, :rating], name: "company_index", length: { type: 10 }, order: { rating: :desc }
     t.index [:firm_id, :type], name: "company_partial_index", where: "(rating > 10)"
+    t.index [:firm_id, :type], name: "company_include_index", include: [:name, :account_id]
```
There's a postgres specific schema in activerecord/test/schema/postgresql_specific_schema.rb which might be a more appropriate file for this change since this won't work on any other adapters.

eileencodes commented on activerecord/test/cases/schema_dumper_test.rb:
```diff
@@ -193,6 +193,15 @@ def test_schema_dumps_partial_indices
+    else
+      assert_equal 't.index ["firm_id", "type"], name: "company_include_index"', index_definition
+    end
+  end
```
I think this test should be moved to activerecord/test//cases/adapters/postgresql/schema_test.rb after moving the schema change below. This feature is postgres specific so it should be contained to postgres files as much as possible.

steve-abrams replied:
Thanks, I've moved the test.

eileencodes commented on activerecord/test/cases/schema_dumper_test.rb:
```diff
@@ -193,6 +193,15 @@ def test_schema_dumps_partial_indices
+    else
+      assert_equal 't.index ["firm_id", "type"], name: "company_include_index"', index_definition
+    end
+  end
```
I think this test should be moved to activerecord/test//cases/adapters/postgresql/schema_test.rb after moving the schema change below. This feature is postgres specific so it should be contained to postgres files as much as possible.

eileencodes reviewed (changes requested):
Congrats on your first PR @steve-abrams! I left some comments about changes we need. The tests look fine other than location, I think the ones I commented on should be moved to the postgres specific tests.

steve-abrams commented on activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb:
```diff
@@ -116,6 +116,10 @@ def supports_partial_index?
 
+      def supports_index_include?
+        false
+      end
```
Good point, I've removed it. I also realized, this option is supported starting in version 11 of Postgres, so I updated the postgresql version of this method to account for that.

steve-abrams commented:
Thanks for the feedback @eileencodes, I've made the suggested changes and I _think_ I've moved everything I can to postgres-specific files. Let me know what you think!

fatkodima commented:
There is an existing PR with this feature https://github.com/rails/rails/pull/37515 from @sebastian-palma

eileencodes commented:
> There is an existing PR with this feature https://github.com/rails/rails/pull/37515 from @sebastian-palma

We can give them credit on this PR. Reopening a PR from 2020 means someone is going to have to fix the conflicts.

That aside, I need someone more familiar with Postgres to review this as I really don't have experience with this adapter/client and consequences of this feature.

eileencodes reviewed (approved):
This looks good to me, just need someone with more postgres experience to offer an opinion on the feature.

steve-abrams commented:
> There is an existing PR with this feature https://github.com/rails/rails/pull/37515 from @sebastian-palma

>> We can give them credit on this PR.

I'm happy to add credit, would I just add @sebastian-palma to the changelog? 

----- 

I also went ahead and squashed, rebased, and fixed up the conflicts for the next reviewer.

eileencodes reviewed (commented):
You would make the suggested change I left in the changelog and then add the following to your commit message:

```
Co-authored-by: Sebastián Palma <vnhnhm@gmail.com>
```

(Got Sebastián's email from Rails commit history, github has no way of finding it in the UI but coauthored doesnt work without it).

sebastian-palma commented:
All yours @steve-abrams thanks for taking the time to push for this, no credits needed. Thanks @eileencodes , @fatkodima for remembering about it

fatkodima commented on activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:
```diff
@@ -799,6 +799,16 @@ def rename_column(table_name, column_name, new_column_name)
+      #
+      # generates:
+      #
+      #   CREATE UNIQUE INDEX index_accounts_on_branch_id ON accounts USING btree(branch_id) INCLUDE (party_id)
+      #
+      # Note: only supported by PostgreSQL with btree indexes.
```
`UNIQUE` is extra.

> with btree indexes

That was in 11 version, now also `GIST` and `SP-GIST` (https://www.postgresql.org/docs/current/sql-createindex.html). I would remove these extra words: `Note: only supported by PostgreSQL`

steve-abrams replied:
Good point. I've updated the wording.

fatkodima replied:
You forgot to remove `UNIQUE.`

fatkodima commented on activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb:
```diff
@@ -24,6 +27,12 @@ def visit_CheckConstraintDefinition(o)
 
+          def visit_CreateIndexDefinition(o)
+            super.dup.tap do |sql|
+              sql << "INCLUDE (#{quoted_include_columns(o.index.include)})" if supports_index_include? && o.index.include
```
I think, this should be better done in the `abstract/schema_creation.rb` base class since SQL Server also [supports](https://www.sqlservertutorial.net/sql-server-indexes/sql-server-indexes-with-included-columns/) include indexes with the same syntax.

steve-abrams replied:
I should have commented this in the PR description. I saw SQL Server supports INCLUDE, but I don't have any experience with SQL Server, so my intent was to implement an initial iteration on Postgres and leave things open for SQL Server in a followup. Not sure how the maintainers like to handle things like this, but I'm a big fan of small iterations for features.

fatkodima replied:
Ah, I see another comment from Eileen regarding this -https://github.com/rails/rails/pull/44803#discussion_r839473577.
Yes, I think this is too postgres specific, better to keep it where it is.

fatkodima commented on activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb:
```diff
@@ -84,6 +93,10 @@ def add_column_options!(sql, options)
             super
           end
 
+          def quoted_include_columns(o)
+            String === o ? o : quoted_include_columns_for_index(o)
+          end
```
Isn't existing `quoted_columns` method already covers this?

steve-abrams replied:
It is similar, but `quoted_columns` is fairly specific to the behavior surrounding the column options. It makes direct use of the `index.columns` and `index.column_options`, whereas for the `include` columns, we just need `index.include`. I considered refactoring `quoted_columns` to be a bit more generic, but the method it calls: `quoted_include_columns_for_index` also behaves slightly differently in how it handles the incoming params.

fatkodima replied:
👍 Yes, missed that it uses other index-specific settings.

fatkodima commented on activerecord/test/schema/postgresql_specific_schema.rb:
```diff
@@ -141,4 +141,8 @@
                                            options: "PARTITION OF measurements FOR VALUES IN (2)")
   end
+
+  if supports_partial_index?
```
`supports_index_include?`?

steve-abrams replied:
Oof, thanks for catching this! I've fixed it.

fatkodima commented on activerecord/test/cases/migration/index_test.rb:
```diff
@@ -256,6 +256,22 @@ def test_add_partial_index
+
+        def test_add_index_with_included_column
+          connection.add_index("testings", "last_name", include: :foo)
+          assert connection.index_exists?("testings", "last_name", include: :foo)
```
`index_exists?` should be extended to also add a check for `include`. 
If you rewrite the first line with `connection.add_index("testings", "last_name")`, the second line will still pass.

steve-abrams replied:
Nice catch! I updated `index_exists?` with an additional check to cover the `include` option.

fatkodima reviewed (commented):
Good job 👍 Added some small comments.

steve-abrams commented:
Thanks for the notes @fatkodima! I responded to your comments and made a few updates based on your suggestions.

steve-abrams commented:
The [build is failing](https://buildkite.com/rails/rails/builds/85785#27814880-0516-4938-af78-ebf4e1311387) and it seems unrelated to the most recent commit 🤔. I verified the failure locally, but reverting the most recent commit still failed locally with seed 20932, so I went ahead and re-pushed to see if it is a flaky failure.

steve-abrams commented:
Hi @eileencodes! I just wanted to check if there was anyone I should ping to keep this moving or if we are just waiting for another reviewer to be available. No hurry, just making sure there isn't anything on my end to be done.

fatkodima commented:
@steve-abrams Could you please rebase?

fatkodima commented:
@yahonda Seems like the author of this PR is not very active (looking at github/twitter activity).
If there would be no activity for some time, would you mind to rebase this yourself, or I can rebase this in another PR? So this very useful feature will go into the new `7.2` version?

steve-abrams commented:
Thanks for the ping @fatkodima! I'm still active, somehow I didn't get any emails or notifications until you pinged me directly. I'll see if I can update my settings so I don't miss things like this again.

@yahonda, I've rebased. The only real change as a result of the rebase was the logic I had added in `ActiveRecord::ConnectionAdapters::SchemaStatements#index_exists?` was moved to [`ActiveRecord::ConnectionAdapters::IndexDefinition#defined_for?`](https://github.com/rails/rails/pull/44803/files#diff-fac852e063eee718c2af44dec6f80b1b46f52f8aae7936e5eea6b2eab08cf2f7R52-R58). Let me know if there's anything else I can do to keep this moving.

yahonda commented on activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb:
```diff
@@ -758,6 +763,15 @@ def foreign_key_column_for(table_name) # :nodoc:
         end
 
+        def quoted_include_columns_for_index(column_names) # :nodoc:
+          return column_names if column_names.is_a?(Symbol)
```
Would you quote `column_names` here?  I wanted every identifier to be quoted. 

`test_include_index` test generates this statement:
```sql
CREATE INDEX "include" ON "ex" ("id")INCLUDE (number)
```

- To be
```sql
CREATE INDEX "include" ON "ex" ("id")INCLUDE ("number")
```

steve-abrams replied:
✅

steve-abrams commented on activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb:
```diff
@@ -27,6 +30,12 @@ def visit_CheckConstraintDefinition(o)
 
+          def visit_CreateIndexDefinition(o)
+            super.dup.tap do |sql|
+              sql << "INCLUDE (#{quoted_include_columns(o.index.include)})" if supports_index_include? && o.index.include
```
✅

steve-abrams commented:
Thanks @yahonda, I've applied the suggested changes and rebased, would you take another look?

yahonda commented:
Looks good. Would you squash your commits?

steve-abrams commented:
Sure thing @yahonda, I've squashed them.

yahonda commented:
Thanks!
